drivers/flash/spi_nor.c

```diff
@@ -398,29 +400,36 @@ static int spi_nor_erase(struct device *dev, off_t addr, size_t size)
 			/* chip erase */
 			spi_nor_cmd_write(dev, SPI_NOR_CMD_CE);
 			size -= params->size;
-		} else if ((size >= SPI_NOR_BLOCK_SIZE)
-			   && SPI_NOR_IS_BLOCK_ALIGNED(addr)) {
-			/* 64 KiB block erase */
-			spi_nor_cmd_addr_write(dev, SPI_NOR_CMD_BE, addr,
-			NULL, 0);
-			addr += SPI_NOR_BLOCK_SIZE;
-			size -= SPI_NOR_BLOCK_SIZE;
-		} else if ((size >= SPI_NOR_BLOCK32_SIZE)
-			   && SPI_NOR_IS_BLOCK32_ALIGNED(addr)) {
-			/* 32 KiB block erase */
-			spi_nor_cmd_addr_write(dev, SPI_NOR_CMD_BE_32K, addr,
-					       NULL, 0);
-			addr += SPI_NOR_BLOCK32_SIZE;
-			size -= SPI_NOR_BLOCK32_SIZE;
-		} else if ((size >= SPI_NOR_SECTOR_SIZE)
-			   && SPI_NOR_IS_SECTOR_ALIGNED(addr)) {
-			/* sector erase */
-			spi_nor_cmd_addr_write(dev, SPI_NOR_CMD_SE, addr,
-					       NULL, 0);
-			addr += SPI_NOR_SECTOR_SIZE;
-			size -= SPI_NOR_SECTOR_SIZE;
+		} else if (erase_types[0].size > 0
```

Comment:
I would prefer to see the erase code loop over the array rather than have a tree of four hard-coded branches.  Could something like this work?

```
if (size == params->size) {
  /* chip erase */
  spi_nor_cmd_write(dev, params->ce_cmd);
  size -= params->size;
} else {
  const struct spi_nor_erase_type *etp = params->erase_types;
  const struct spi_nor_erase_type *etpe = etp + ARRAY_SIZE(params->erase_types);

  while ((etp < etpe) && (etp->size != 0)) {
      if ((size >= etp->size) && SPI_NOR_IS_ALIGNED(addr, etp->size)) {
         spi_nor_cmd_addr_write(dev, etp->op_code, addr, NULL, 0);
	 addr += etp->size;
	 size -= etp->size;
	 break;
      }
      ++etp;
  }
  if ((etp == etpe) || (etp->size == 0)) {
      /* too small or misaligned */
      retval = -EINVAL;
      goto out;
  }
}
```

Reply:
Much cleaner solution :) Done.